File: android/src/main/kotlin/com/swmansion/detour/DetourFlutterPlugin.kt
package com.swmansion.detour

import android.app.Activity
import android.content.Intent
import android.net.Uri
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import com.swmansion.detour.analytics.DetourAnalytics
import com.swmansion.detour.analytics.DetourEventNames
import com.swmansion.detour.models.LinkProcessingMode
import com.swmansion.detour.models.LinkResult

class DetourFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    EventChannel.StreamHandler,
    PluginRegistry.NewIntentListener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    private var isConfigured = false

    private val scope = CoroutineScope(Dispatchers.Main.immediate + SupervisorJob())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "detour_flutter_plugin")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "detour_flutter_plugin/links")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addOnNewIntentListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding?.removeOnNewIntentListener(this)
        activity = null
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addOnNewIntentListener(this)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeOnNewIntentListener(this)
        activity = null
        activityBinding = null
    }

    override fun onNewIntent(intent: Intent): Boolean {
        if (!isConfigured) return false
        val sink = eventSink ?: return false

        scope.launch {
            when (val nativeResult = Detour.processLink(intent)) {
                is LinkResult.Error -> {
                    sink.error("NATIVE_ERROR", nativeResult.exception.message, null)
                }
                else -> sink.success(toFlutterMap(nativeResult))
            }
        }
        return true
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val args = call.arguments as? Map<*, *>
                val apiKey = args?.get("apiKey") as? String
                val appID = args?.get("appID") as? String
                if (apiKey.isNullOrBlank() || appID.isNullOrBlank()) {
                    result.error("INVALID_ARGS", "apiKey and appID are required", null)
                    return
                }

                val modeRaw = (args["linkProcessingMode"] as? String) ?: "all"
                val mode = when (modeRaw) {
                    "web-only" -> LinkProcessingMode.WEB_ONLY
                    "deferred-only" -> LinkProcessingMode.DEFERRED_ONLY
                    else -> LinkProcessingMode.ALL
                }

                val appContext = activity?.applicationContext
                if (appContext == null) {
                    result.error("NO_ACTIVITY", "Plugin requires an attached Activity to configure", null)
                    return
                }

                val config = DetourConfig(
                    apiKey = apiKey,
                    appId = appID,
                    linkProcessingMode = mode,
                    storage = null,
                )

                Detour.initialize(appContext, config)
                isConfigured = true
                result.success(null)
            }

            "resolveInitialLink" -> {
                if (!isConfigured) {
                    result.error("NOT_CONFIGURED", "Call configure() first", null)
                    return
                }

                val initialIntent = activity?.intent ?: Intent()
                scope.launch {
                    when (val nativeResult = Detour.processLink(initialIntent)) {
                        is LinkResult.Error -> {
                            result.error("NATIVE_ERROR", nativeResult.exception.message, null)
                        }
                        else -> result.success(toFlutterMap(nativeResult))
                    }
                }
            }

            "processLink" -> {
                if (!isConfigured) {
                    result.error("NOT_CONFIGURED", "Call configure() first", null)
                    return
                }

                val args = call.arguments as? Map<*, *>
                val rawUrl = args?.get("url") as? String
                if (rawUrl.isNullOrBlank()) {
                    result.error("INVALID_ARGS", "url is required", null)
                    return
                }

                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(rawUrl))
                scope.launch {
                    when (val nativeResult = Detour.processLink(intent)) {
                        is LinkResult.Error -> {
                            result.error("NATIVE_ERROR", nativeResult.exception.message, null)
                        }
                        else -> result.success(toFlutterMap(nativeResult))
                    }
                }
            }

            "logEvent" -> {
                if (!isConfigured) {
                    result.error("NOT_CONFIGURED", "Call configure() first", null)
                    return
                }

                val args = call.arguments as? Map<*, *>
                val eventName = args?.get("eventName") as? String
                if (eventName.isNullOrBlank()) {
                    result.error("INVALID_ARGS", "eventName is required", null)
                    return
                }

                val enumValue = DetourEventNames.entries.firstOrNull { it.eventName == eventName }
                if (enumValue == null) {
                    result.error(
                        "UNSUPPORTED_EVENT",
                        "Android SDK supports predefined DetourEventNames only. Received: $eventName",
                        null,
                    )
                    return
                }

                val data = args["data"]
                DetourAnalytics.logEvent(enumValue, data)
                result.success(null)
            }

            "logRetention" -> {
                if (!isConfigured) {
                    result.error("NOT_CONFIGURED", "Call configure() first", null)
                    return
                }

                val args = call.arguments as? Map<*, *>
                val eventName = args?.get("eventName") as? String
                if (eventName.isNullOrBlank()) {
                    result.error("INVALID_ARGS", "eventName is required", null)
                    return
                }

                DetourAnalytics.logRetention(eventName)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun toFlutterMap(nativeResult: LinkResult): Map<String, Any?> {
        return when (nativeResult) {
            is LinkResult.Success -> {
                mapOf(
                    "processed" to true,
                    "link" to mapOf(
                        "url" to nativeResult.url,
                        "route" to nativeResult.route,
                        "pathname" to nativeResult.pathname,
                        "params" to nativeResult.params,
                        "type" to when (nativeResult.type) {
                            com.swmansion.detour.models.LinkType.DEFERRED -> "deferred"
                            com.swmansion.detour.models.LinkType.VERIFIED -> "verified"
                            com.swmansion.detour.models.LinkType.SCHEME -> "scheme"
                        },
                    ),
                )
            }
            is LinkResult.NoLink,
            is LinkResult.NotFirstLaunch,
            is LinkResult.Error -> {
                mapOf(
                    "processed" to true,
                    "link" to null,
                )
            }
        }
    }
}

File: ios/Classes/DetourFlutterPlugin.swift
import Flutter
import UIKit
import Detour

public class DetourFlutterPlugin: NSObject, FlutterPlugin, FlutterApplicationLifeCycleDelegate, FlutterSceneLifeCycleDelegate {
    private var config: DetourConfig?
    private var launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    private var eventSink: FlutterEventSink?
    private var pendingRuntimeURLs: [URL] = []

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "detour_flutter_plugin",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "detour_flutter_plugin/links",
            binaryMessenger: registrar.messenger()
        )
        let instance = DetourFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        registrar.addApplicationDelegate(instance)
        // Register for scene lifecycle callbacks when available (iOS 13+).
        // Use selector-based invocation for backwards compatibility with older Flutter SDKs.
        if let registrarObject = registrar as? NSObject {
            let selector = NSSelectorFromString("addSceneDelegate:")
            if registrarObject.responds(to: selector) {
                _ = registrarObject.perform(selector, with: instance)
            }
        }
    }

    // MARK: - FlutterApplicationLifeCycleDelegate

    public func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any] = [:]
    ) -> Bool {
        self.launchOptions = launchOptions
        return true
    }

    public func application(
        _ app: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        guard shouldHandleSchemeRuntimeLink(url) else { return false }
        emitOrQueueRuntimeURL(url)
        return true
    }

    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb else { return false }
        guard let url = userActivity.webpageURL else { return false }
        guard shouldHandleWebRuntimeLink(url) else { return false }

        emitOrQueueRuntimeURL(url)
        return true
    }

    // MARK: - FlutterSceneLifeCycleDelegate

    @available(iOS 13.0, *)
    public func scene(
        _ scene: UIScene,
        willConnectTo session: UISceneSession,
        options connectionOptions: UIScene.ConnectionOptions?
    ) -> Bool {
        guard let connectionOptions else { return false }

        var handled = false

        for urlContext in connectionOptions.urlContexts {
            if handleRuntimeURLCandidate(urlContext.url) {
                handled = true
            }
        }

        for userActivity in connectionOptions.userActivities {
            guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
                  let url = userActivity.webpageURL,
                  shouldHandleWebRuntimeLink(url) else {
                continue
            }
            emitOrQueueRuntimeURL(url)
            handled = true
        }

        return handled
    }

    @available(iOS 13.0, *)
    public func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) -> Bool {
        var handled = false

        for urlContext in URLContexts {
            if handleRuntimeURLCandidate(urlContext.url) {
                handled = true
            }
        }

        return handled
    }

    @available(iOS 13.0, *)
    public func scene(_ scene: UIScene, continue userActivity: NSUserActivity) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb else { return false }
        guard let url = userActivity.webpageURL else { return false }
        guard shouldHandleWebRuntimeLink(url) else { return false }

        emitOrQueueRuntimeURL(url)
        return true
    }

    // MARK: - FlutterMethodCallHandler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            guard let args = call.arguments as? [String: Any],
                  let config = configFromArgs(args) else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing apiKey or appID", details: nil))
                return
            }
            self.config = config
            flushPendingRuntimeURLsIfPossible()
            result(nil)

        case "resolveInitialLink":
            guard let config = self.config else {
                result(FlutterError(code: "NOT_CONFIGURED", message: "Call configure() first", details: nil))
                return
            }
            let launchOpts = self.launchOptions
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                Detour.shared.resolveInitialLink(config: config, launchOptions: launchOpts) { r in
                    result(self.detourResultToMap(r))
                }
            }

        case "processLink":
            guard let config = self.config else {
                result(FlutterError(code: "NOT_CONFIGURED", message: "Call configure() first", details: nil))
                return
            }
            guard let args = call.arguments as? [String: Any],
                  let url = args["url"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "url is required", details: nil))
                return
            }
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                let r = await Detour.shared.processLink(url, config: config)
                result(self.detourResultToMap(r))
            }

        case "logEvent":
            guard let args = call.arguments as? [String: Any],
                  let eventName = args["eventName"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "eventName is required", details: nil))
                return
            }
            let data = args["data"] as? [String: Any]
            guard let typedEventName = DetourEventName(rawValue: eventName) else {
                result(FlutterError(code: "UNSUPPORTED_EVENT", message: "Only predefined DetourEventName values are supported", details: nil))
                return
            }
            Task { @MainActor in
                DetourAnalytics.logEvent(typedEventName, data: data)
            }
            result(nil)

        case "logRetention":
            guard let args = call.arguments as? [String: Any],
                  let eventName = args["eventName"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "eventName is required", details: nil))
                return
            }
            Task { @MainActor in
                DetourAnalytics.logRetention(eventName)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Helpers

    private func detourResultToMap(_ r: DetourResult) -> [String: Any?] {
        var map: [String: Any?] = ["processed": r.processed]
        if let link = r.link {
            map["link"] = [
                "url": link.url,
                "route": link.route,
                "pathname": link.pathname,
                "params": link.params,
                "type": link.type.rawValue,
            ] as [String: Any]
        } else {
            map["link"] = nil
        }
        return map
    }

    private func configFromArgs(_ args: [String: Any]) -> DetourConfig? {
        guard let apiKey = args["apiKey"] as? String,
              let appID = args["appID"] as? String else { return nil }
        let shouldUseClipboard = args["shouldUseClipboard"] as? Bool ?? true
        let modeStr = args["linkProcessingMode"] as? String ?? "all"
        let mode: LinkProcessingMode
        switch modeStr {
        case "web-only": mode = .webOnly
        case "deferred-only": mode = .deferredOnly
        default: mode = .all
        }
        return DetourConfig(
            apiKey: apiKey,
            appID: appID,
            shouldUseClipboard: shouldUseClipboard,
            linkProcessingMode: mode
        )
    }

    private func shouldHandleSchemeRuntimeLink(_ url: URL) -> Bool {
        guard !isWebURL(url) else { return false }
        guard !isInfrastructureURL(url) else { return false }
        guard config?.linkProcessingMode != .webOnly else { return false }
        guard config?.linkProcessingMode != .deferredOnly else { return false }
        return true
    }

    private func shouldHandleWebRuntimeLink(_ url: URL) -> Bool {
        guard isWebURL(url) else { return false }
        guard !isInfrastructureURL(url) else { return false }
        guard config?.linkProcessingMode != .deferredOnly else { return false }
        return true
    }

    private func handleRuntimeURLCandidate(_ url: URL) -> Bool {
        if shouldHandleWebRuntimeLink(url) || shouldHandleSchemeRuntimeLink(url) {
            emitOrQueueRuntimeURL(url)
            return true
        }
        return false
    }

    private func canProcessRuntimeURL(_ url: URL, with config: DetourConfig) -> Bool {
        if isInfrastructureURL(url) {
            return false
        }

        switch config.linkProcessingMode {
        case .all:
            return true
        case .webOnly:
            return isWebURL(url)
        case .deferredOnly:
            return false
        }
    }

    private func isWebURL(_ url: URL) -> Bool {
        guard let scheme = url.scheme?.lowercased() else { return false }
        return scheme == "http" || scheme == "https"
    }

    private func isInfrastructureURL(_ url: URL) -> Bool {
        let raw = url.absoluteString
        return raw.isEmpty || raw == "about:blank"
    }

    private func emitOrQueueRuntimeURL(_ url: URL) {
        guard let config = self.config, let sink = self.eventSink else {
            pendingRuntimeURLs.append(url)
            return
        }

        guard canProcessRuntimeURL(url, with: config) else {
            return
        }

        Task { @MainActor [weak self] in
            guard let self = self else { return }
            let result = await Detour.shared.processLink(url, config: config)
            sink(self.detourResultToMap(result))
        }
    }

    private func flushPendingRuntimeURLsIfPossible() {
        guard let config = self.config, let sink = self.eventSink else { return }
        guard !pendingRuntimeURLs.isEmpty else { return }

        let queued = pendingRuntimeURLs
        pendingRuntimeURLs.removeAll()

        for url in queued {
            guard canProcessRuntimeURL(url, with: config) else { continue }

            Task { @MainActor [weak self] in
                guard let self = self else { return }
                let result = await Detour.shared.processLink(url, config: config)
                sink(self.detourResultToMap(result))
            }
        }
    }
}

// MARK: - FlutterStreamHandler

extension DetourFlutterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        flushPendingRuntimeURLsIfPossible()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
